core/src/main/scala/chisel3/internal/firrtl/Converter.scala

```diff
@@ -298,10 +329,24 @@ private[chisel3] object Converter {
     case d => d.specifiedDirection
   }
 
-  def extractType(data: Data, info: SourceInfo): fir.Type = extractType(data, false, info)
+  def extractType(data: Data, info: SourceInfo): fir.Type = extractType(data, false, info, true, true)
 
-  def extractType(data: Data, clearDir: Boolean, info: SourceInfo, checkConst: Boolean = true): fir.Type = data match {
-    case _ if (checkConst && data.isConst) => fir.ConstType(extractType(data, clearDir, info, false))
+  def extractType(
+    data:       Data,
+    clearDir:   Boolean,
+    info:       SourceInfo,
+    checkProbe: Boolean,
+    checkConst: Boolean
+  ): fir.Type = data match {
+    // extract underlying type for probe
+    case _ if (checkProbe && data.probeInfo.nonEmpty) =>
+      if (data.probeInfo.get.writable) {
+        fir.RWProbeType(extractType(data, clearDir, info, false, checkConst))
+      } else {
+        fir.ProbeType(extractType(data, clearDir, info, false, checkConst))
+      }
+    // extract underlying type for const
+    case _ if (checkConst && data.isConst) => fir.ConstType(extractType(data, clearDir, info, checkProbe, false))
```

Comment:
```suggestion
    case d if (checkProbe && d.probeInfo.nonEmpty) =>
      if (d.probeInfo.get.writable) {
        fir.RWProbeType(extractType(d, clearDir, info, false, checkConst))
      } else {
        fir.ProbeType(extractType(d, clearDir, info, false, checkConst))
      }
    // extract underlying type for const
    case d if (checkConst && d.isConst) => fir.ConstType(extractType(d, clearDir, info, checkProbe, false))
```
It's generally better to bind the variable and use that rather than use the argument to the `match`. It is "defensive programming" against a later refactor that changes the code above such that `data` is no longer the argument to the match (eg. if it became `data.cloneType` or something)`.